nengo/cache.py

```diff
@@ -321,6 +321,11 @@ def shrink(self, limit=None):
             Maximum size of the cache in bytes.
         """
         if self.readonly:
+            warnings.warn("Cannot shrink a readonly cache.")
+            return
+
+        if self._index is None:
```

Comment:
`self._index` won't be reset to `None` in `__exit__`. Maybe we should do that for consistency? (I think technically `shrink` would work outside of the `with` block if there has been one call to `__enter__` before even after the `__exit__`.)